experimental/Geometry/Schemes.jl

```diff
@@ -0,0 +1,1244 @@
+
+  function AffSchMorphism( domain::AffineScheme{S,Td,Ud},
+                           codomain::AffineScheme{S,Tc,Uc}, pullback::AlgHom
+                           ) where {S,Td,Ud,Tc,Uc}
+
+   if base_ring(domain) != base_ring(codomain)
+      error( "the base rings of the domain and the codomain do not coincide!" )
+    end
+    k = base_ring(domain)
+    #if domain(pullback) != ambient_ring(R) || codomain(pullback) != domain.R
+      #error( "the domain and codomain of the given ring homomorphism is not compatible with the affine schemes." )
+    #end
+    x = new{S,Td,Ud,Tc,Uc}()
+    x.domain = domain
+    x.codomain = codomain
+    x.pullback = pullback
+    # TODO: Implement the checks for the homomorphism to be well defined (on demand).
+    return x
+
+  end
+
+  function AffSchMorphism( domain::AffineScheme{S,Td,Ud},
+      codomain::AffineScheme{S,Tc,Uc}, imgs_frac::Vector{Frac{Uc}}
+                           ) where {S,Td,Ud,Tc,Uc}
+    if base_ring(domain) != base_ring(codomain)
+      error( "the base rings of the domain and the codomain do not coincide!" )
+    end
+    k = base_ring(domain)
+    #if domain(pullback) != codomain.R || codomain(pullback) != domain.R
+      #error( "the domain and codomain of the given ring homomorphism is not compatible with the affine schemes." )
+    #end
+    x = new{S,Td,Ud,Tc,Uc}()
+    x.domain = domain
+    x.codomain = codomain
+    x.imgs_frac = imgs_frac
+    # TODO: Implement the checks for the homomorphism to be well defined (on demand).
+    return x
+  end
+end
+
+domain(f::AffSchMorphism) = f.domain
+codomain(f::AffSchMorphism) = f.codomain
+
+# Construct the pullback on the level of coordinate rings 
+# from the fractional representation 
+function pullback(f::AffSchMorphism)
+  if isdefined(f, :pullback)
+    return f.pullback
+  end
+  if !isdefined(f, :imgs_frac )
+    error( "Neither the fractional representation, nor the pullback is defined for this morphism." )
+  end
+  S = ambient_ring(codomain(f))
+  T = ambient_ring(domain(f))
+  R = ambient_ring(root(codomain(f)))
+  # TODO reconstruct the ring homomorphism S -> T from the fractional representation 
+  # and describe it in terms of R -> T with R = ambient_ring(root(codomain(f)))
+
+  n = length( f.imgs_frac )
+  if n != length( gens( R ))
+    error( "Number of variables in the ambient ring of the root does not coincide with the number of images provided for the homomorphism." )
+  end
+  for i in (1:n)
+    p = numerator( f.imgs_frac[i] )
+    q = denominator( f.imgs_frac[i] )
+    u = prod( denoms( domain(f)))
+    (k,a) = coeffs_in_radical( q, u )
+    
+  end
+  #TODO: Finish this once we know how to deal with the radical membership in Oscar.
+
+end
+
+# Construct the fractional representation of the morphism 
+# from the explicit algebra homomorphism. 
+function imgs_frac(f::AffSchMorphism)
+  # first check if this variable is already cached
+  if isdefined(f, Symbol("imgs_frac"))
```

Comment:
```suggestion
  if isdefined(f, :imgs_frac))
```

Reply:
I already went through the code to catch these. But some of them seem to have slipped my notice...